### PR TITLE
More generic adapter

### PR DESCRIPTION
**Pull request by dexterlohnes**
Made the adapter class more generic to better handle Slack development. Big changes were:

1) Making the main overridden functions in adapter be explicit about giving feedback to user
2) Took functions out of util class and put them in redditAdapter. **I think we'll probably want to create an abstract extractor class moving forward but I have yet to do this**
3) Made all "onDeposit/on..." type functions in abstract Adapter class async by default to reflect their implementations better.

All tests should be passing, but I'd appreciate your double checking everything especially in prod (I can't verify the bot myself atm).

### CODE REVIEW

File: src/adapters/abstractAdapter.js [hunk 1]
```diff
@@ -25,12 +25,34 @@ class Adapter {
 
   }
 
-  onTransfer (sourceAccount, targetAccount, amount) {
-    // Hook function
+  async onTransfer (sourceAccount, targetAccount, amount) {
+      this.sendTransferConfirmation(sourceAccount, amount)
```
Comment (shredding) [inline src/adapters/abstractAdapter.js]: What's the idea behind this? onTransfer and sendTransferConfirmation do exactly the same thing (whereas onTransfer is more generic).

File: src/adapters/abstractAdapter.js [hunk 2]
```diff
@@ -25,12 +25,34 @@ class Adapter {
```
Comment (shredding) [inline src/adapters/abstractAdapter.js]: can you rename `abstractAdapter.js` to `abstract-adapter.js` case sensitive file names are a pain in the ass on some OS.

File: src/stellar/index.js [hunk 3]
```diff
@@ -22,6 +22,8 @@ module.exports = async function (models) {
     callBuilder.cursor(latestTx.cursor)
   }
 
+  // What happens here if latestTx is null for any reason?
```
Comment (shredding) [inline src/stellar/index.js]: Do not add questions in the source :-)

If it's null we start at the first incoming entry of the account, which is a good idea because null would happen if we haven't collect anything yet.

File: src/adapters/reddit.js [hunk 4]
```diff
@@ -55,7 +57,32 @@ function formatMessage(txt) {
...
-  async onDeposit (sourceAccount, amount) {
+  extractTipAmount (tipText) {
+    const matches =  tipText.match(/\+\+\+([\d\.]*)[\s{1}]?XLM/i)
+    if (matches) {
+        return new Big(matches[1])
+    }
+    return undefined
+  }
+
+  extractWithdrawal (body) {
```
Comment (shredding) [inline src/adapters/reddit.js]: Is it a good idea to have a hook func for this in the abstract or not? I'm not sure - do u need it on slack as well?

File: test/test.adapter.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
 const assert = require('assert')
-const Adapter = require('../src/adapters/abstract')
+const Adapter = require('../src/adapters/abstractAdapter')
```
Comment (shredding) [inline test/test.adapter.js]: rename -> abstract-adapter

File: src/adapters/abstractAdapter.js (hunk 1 above)
Comment (dexterlohnes) [reply src/adapters/abstractAdapter.js]: My thinking is that 100% of all adapters should be sending confirmations; as such, it is appropriate to create a specific overridable function to handle this behavior. It's much clearer code for anyone who's wanting to create their own integration / adapter. 

Likewise, it's unclear if all or any integrations are going to want to perform additional functionality "onTransfer". Thus, I believe it makes sense to leave this is as being capable of being overridden, but not making it necessary. 

In short, my thinking is that if most "onTransfer" functions actually just send a transfer confirmation, it makes sense for the function to be called "sendTransferConfirmation" :)

File: src/adapters/abstractAdapter.js (hunk 2 above)
Comment (dexterlohnes) [reply src/adapters/abstractAdapter.js]: Will do.

File: src/stellar/index.js (hunk 3 above)
Comment (dexterlohnes) [reply src/stellar/index.js]: No problem. What's the best way for me to add questions like this? Something through github or just contact through Slack?

File: src/adapters/reddit.js (hunk 4 above)
Comment (dexterlohnes) [reply src/adapters/reddit.js]: Each platform should allow for users to make withdrawals, but each one may be different from the last. I think it makes sense to have extractWithdrawal() as part of some interface, but as per the issue I created states I'm not convinced the abstract-adapter is the place for it. Might make sense to have an extractor class corresponding to each platform as well.

File: test/test.adapter.js (hunk 5 above)
Comment (dexterlohnes) [reply test/test.adapter.js]: 👍

**Comment by dexterlohnes** [comment]
Everything concrete in your feedback has been addressed. Going to move on to resolving the conflicts now.

**Comment by dexterlohnes** [comment]
Reverted the extractor interface commits, so everything is ready to merge. Conflicts resolved.

File: src/adapters/abstract-adapter.js [hunk 6]
```diff
@@ -73,6 +73,27 @@ class Adapter extends EventEmitter {
...
+  async sendDepositConfirmation (sourceAccount, amount) {
+      // Override me
+  }
+
+  async sendTransferConfirmation (sourceAccount, amount) {
+      // Override me
+  }
+
+    /**
+     * // Being that each platform (Reddit, Twitter, Slack...) can have different
```
Comment (shredding) [inline src/adapters/abstract-adapter.js:85]: Please do not comment out code, remove it. That's why we have version control.

**Comment by shredding** [comment]
There are two things to commit here, right?

1. Changing `onDeposit` to `sendDepositConfirmation` and  change `onTransfer` to `sendTransferConfirmation`

2. Rename `abstract.js` to `abstract-adapter.js`

I have reservations towards that and here's why:

Regarding `1.`:

The idea of the abstract adapter is to decouple the `bot-system` (that is the one managing balances & communicating with stellar) with `platforms` (that is somewhere where people can tip like twitter, reddit and so on).

Decoupling means that both systems do know as little about each other as possible and make as few assumptions about each other as possible, hence the interface should be the minimum required set of things required to allow for as many things possible.

When we rename `onDeposit` to `sendDepositConfirmation`, we break with this idea for quite a lot of reasons:

- if people do want to send a confirmation but as well do something else (broadcasting the event, write to logs, whatever), the method naming is counterintuitive. This is because the name is too `concrete`!

- if people do not want to send a confirmation (e.g. because the user has opted out, the platform does not have a messaging system or the author wants to support multisignature signing and requires more information before she can confirm), the method name is as well too `concrete` and would prevent the author from doing what she wants to do (or require them to be hacky and hijack the `sendConfirmation` method to do what they want.

- if we'd rename this method, we'd need to rename the various error options as well, which would become a very bloated interface.

- note that the hook methods do as well emit events as an option, if people want to go with the event driven architecture that javascript recommends. Events are something that happens, and not something that instructs.

In the end we would end up with an interface where the `bot-system` and the `platform` where not decoupled but where the bot-system would tell the other system what to do. That is overly bloated and an unnecessary restriction towards possible implementations.

Regarding `2.`:

The file is already lying in a folder named `adapter` and the reason for this structure is that this folder contains adapter. Why adding that to the filename when it's already obvious by the overall structure?

We do not have `reddit-adapter.js` and we do not have `account-model.js` or `transaction-model.js`, because it's obvious to the filestructure.

To sum up: I have started with [Documentation](https://github.com/shredding/stellar-bot/wiki) and a [how-to on creating adapters](https://github.com/shredding/stellar-bot/wiki/How-To:-Create-an-adapter) and I think it would be a better to add a `structure` section there, explaining what to do and a `best practises` section in the guide on how we recommend adapters to be written.

**Comment by dexterlohnes** [comment]
@shredding 

Regarding point 1, I think we can have both onDeposit and sendDepositConfirmation in the interface and it'd still be a fine implementation (I didn't intended to completely do without 'onDeposit'; sorry if that was ever unclear). But if we're going this way, that's fine. Per my comments yesterday, it's not worth arguing much over once we've both said our piece. 

Point 2 sounds good.
